stan/math/prim/arr/fun/matrix_gpu.hpp

```diff
@@ -25,169 +20,195 @@ enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
-            if (dst.cols() != src.cols) {
-              std::cout << "Copy (GPU -> Eigen): Eigen and stanmathCL matrix_gpu col lengths do no match!" <<
-               std::endl;
-            }
-            int rows = dst.rows();
-            int cols = dst.cols();
+
+    /**
+     * Copies the source matrix that is stored
+     * on the GPU to the destination Eigen 
+     * matrix. 
+     * 
+     * @tparam T type of data in the Eigen matrix
+     * @param src source matrix on the GPU
+     * @param dst destination Eigen matrix
+     * 
+     * @throw <code>std::invalid_argument</code> if the 
+     * matrices do not have matching dimensions
+     * 
+     */
+    template <typename T, int R, int C>
+    void copy(matrix_gpu & src,
+     Eigen::Matrix<T, R, C> & dst) {
+            check_size_match("copy (GPU -> Eigen)",
+             "src.rows()", src.rows(), "dst.rows()", dst.rows());
+            check_size_match("copy (GPU -> Eigen)",
+             "src.cols()", src.cols(), "dst.cols()", dst.cols());
             try {
-              double*  Btemp =  new double[rows * cols];
-              cl::Context ctx = stan::math::get_context();
-              cl::CommandQueue queue = stan::math::get_queue();
-              cl::Buffer buffer = src.buffer();
-              queue.enqueueReadBuffer(buffer,  CL_TRUE,  0,
-                sizeof(double) * dst.size(),  Btemp);
-              for(int j = 0; j < cols; j++) {
-                for(int i = 0; i < rows; i++) {
-                  dst.coeffRef(i, j) = Btemp[i * rows + j];
-                }
-              }
-              delete[] Btemp;
-            } catch (cl::Error& e) {
-              check_ocl_error(e);
+              cl::CommandQueue queue = get_queue();
+              queue.enqueueReadBuffer(src.buffer(),  CL_TRUE,  0,
+                sizeof(T) * dst.size(),  dst.data());
+            } catch (const cl::Error& e) {
+              check_ocl_error("copy GPU->Eigen", e);
             }
     }
 
-    void copy(stan::math::matrix_gpu & src,  stan::math::matrix_gpu & dst) {
-      if (src.rows != dst.rows || src.cols != dst.cols) {
-        app_error("The dimensions of the input and output matrices should match.");
-      }
-      cl::Kernel kernel = stan::math::get_kernel("copy");
-      cl::CommandQueue cmdQueue = stan::math::get_queue();
+    /**
+     * Copies the source matrix to the 
+     * destination matrix. Both matrices 
+     * are stored on the GPU.
+     * 
+     * @param src source matrix
+     * @param dst destination matrix
+     * 
+     * @throw <code>std::invalid_argument</code> if the 
+     * matrices do not have matching dimensions
+     * 
+     */
+    void copy(matrix_gpu& src,  matrix_gpu& dst) { // NOLINT
+      check_size_match("copy (GPU -> GPU)",
+        "src.rows()", src.rows(), "dst.rows()", dst.rows());
+      check_size_match("copy (GPU -> GPU)",
+        "src.cols()", src.cols(), "dst.cols()", dst.cols());
+      cl::Kernel kernel = get_kernel("copy");
```

Comment:
Is this outside of the try/catch because it can't throw?

Reply:
The try/catch scopes scopes were a bit too big (included stuff that either did not throw or has its own try/catch for OpenCL code. This was addressed in the latest commit.